python/ray/util/tracing.py

```diff
@@ -0,0 +1,400 @@
+
+try:
+    from opentelemetry import context, trace
+    from opentelemetry.context.context import Context
+    # from opentelemetry.util.types import Attributes
+    from opentelemetry import propagators
+    from opentelemetry.trace.propagation.textmap import DictGetter
+except ImportError:
+    if os.getenv("RAY_TRACING_ENABLED", "False").lower() in ["true", "1"]:
+        raise Exception(
+            "Install opentelemetry with "
+            "'pip install opentelemetry-api==1.0.0rc1' and "
+            "'pip install opentelemetry-sdk==1.0.0rc1' to enable tracing.")
+
+_nameable = Union[str, Callable[..., Any]]
+_global_is_tracing_enabled = None
+
+
+def is_tracing_enabled() -> bool:
+    """Checks environment variable feature flag to see if tracing is turned on.
+    Tracing is off by default."""
+    global _global_is_tracing_enabled
+    if _global_is_tracing_enabled is None:
+        _global_is_tracing_enabled = os.getenv(
+            "RAY_TRACING_ENABLED", "False").lower() in ["true", "1"]
+    return _global_is_tracing_enabled
+
+
+class DictPropagator:
+    def inject_current_context() -> Dict[Any, Any]:
+        """Inject trace context into otel propagator."""
+        context_dict: Dict[Any, Any] = {}
+        propagators.inject(dict.__setitem__, context_dict)
+        return context_dict
+
+    def extract(context_dict: Dict[Any, Any]):
+        """Given a trace context, extract as a Context."""
+        return cast(Context, propagators.extract(DictGetter(), context_dict))
+
+
+@contextmanager
+def use_context(parent_context) -> Generator[None, None, None]:
+    """Uses the Ray trace context for the span."""
+    new_context = parent_context if parent_context is not None else Context()
+    token = context.attach(new_context)
+    try:
+        yield
+    finally:
+        context.detach(token)
+
+
+def _function_hydrate_span_args(func: Callable[..., Any]):
+    """Get the Attributes of the function that will be reported as attributes
+    in the trace."""
+    runtime_context = get_runtime_context().get()
+
+    span_args = {
+        "ray.remote": "function",
+        "ray.function": func,
+        "ray.pid": str(os.getpid()),
+        "ray.job_id": runtime_context["job_id"].hex(),
+        "ray.node_id": runtime_context["node_id"].hex(),
+    }
+
+    # We only get task ID for workers
+    if ray.worker.global_worker.mode == ray.worker.WORKER_MODE:
+        task_id = (runtime_context["task_id"].hex()
+                   if runtime_context["task_id"] else None)
+        if task_id:
+            span_args["ray.task_id"] = task_id
+
+    worker_id = getattr(ray.worker.global_worker, "worker_id", None)
+    if worker_id:
+        span_args["ray.worker_id"] = worker_id.hex()
+
+    return span_args
+
+
+def _function_span_producer_name(func: Callable[..., Any]) -> str:
+    """Returns the function span name that has span kind of producer."""
+    args = _function_hydrate_span_args(func)
+    assert args is not None
```

Comment:
this doesn't look like a meaningful `assert` -- the called function never returns `None`